### ISSUE TITLE
Showing science station on main screen gives an empty shell

### ISSUE THREAD

**Issue by MCMic** [smcameron/space-nerds-in-space#216]
I already had this problem before, showing the science station on the main screen when in details tab does not work as intended: The labels and frames are there but the information and 3D model are missing.

Maybe related to #215 ?

**Comment by smcameron** [comment]
This is likely due to the selected object being de-selected due to being out of range or low power on science.  It should probably be clearer about that.

**Comment by smcameron** [comment]
This might help somewhat with knowing when nothing is selected: 9c6f39026ea6ec7f31be6e28b8f328023588a579

However, I'm not quite sure I'm understanding the problem: showing science on main screen.  So I presume you are pressing ctrl-O on the science screen to project it to any clients which have the the MAIN SCREEN role (selected on the first login screen when starting the client).

This seems to work for me, but I'm not sure that I'm doing precisely what you're doing to provoke the problem.

**Comment by MCMic** [comment]
Finally got it!

This only happens if science role is not checked on the main screen. I guess this is why it does not trigger in your tests.

So, in our setup I only check one station on each computer, except the server which has main screen, demon screen, TTS and sound server. But it does not have science station and I think this is why it appears as if nothing is selected.

Could #215 be due to the same thing? (the navigation station does not have science station checked)

**Comment by smcameron** [comment]
> This only happens if science role is not checked on the main screen.

Ok, that kind of makes sense.  The way that projecting a station to the main screen works is for snis_server to simply ask any snis_clients that have the main screen role to switch to the requested station.  If the main screen does not have the role for that station checked, I'm actually somewhat surprised it does anything (surprised that it doesn't remain on the main view).

I usually run with only one station having MAIN SCREEN, TEXT TO SPEECH and SOUND SERVER but every other station having every other role (unless somebody's running the limited client due to ancient graphics hardware or natively on Mac... in which case I restrict to engineering, damage control, and science). I should probably try out more restrictive setups more, and also make it clearer what works best (though, I'm not quite sure what works best... maybe it depends on what you like.  If you do not want people switching stations, then you might want to have the roles more locked down.)

I don't think #215 is the same thing.  Even if a station doesn't have the science role, it should still know which (if any) item is selected by science.  I think this is caused by no item being selected on science.  It may be that somehow the selected item is getting de-selected for some non-obvious (or even wrong) reason.

Thanks for the extra info, I'll play around with it.

**Comment by smcameron** [comment]
Ok, I tried it out with two clients, one with MAIN SCREEN, TTS, SOUND, and DEMON, and one with SCIENCE, TTS, SOUND, and DEMON.  When I press Ctrl-O on the SCIENCE screen, the other client switches to the science screen even though it doesn't have the science role.  And since it doesn't have the science role (or nav or weapons) it got the science selection information  This has other implications for other screens too.  Probably having MAIN SCREEN role should imply having the other roles too -- if you have main screen role, you should be forced to have the other roles as well.

**Comment by MCMic** [comment]
Yeah I agree having main role should imply all the roles that can be displayed on it.

**Comment by smcameron** [comment]
Ok, this should fix it: 866338ff5e62ba3f8b563c6167c9c73965c5a4b3

There is still the minor problem that selecting the MAIN SCREEN role on the lobby screen does not automatically fill in the checkboxes for the other role, but they are set, so it's just a visual bug, it's still doing the right thing.  I'll fix that later (maybe).

**Comment by smcameron** [comment]
This makes the UI show the other stations selected if MAIN is selected.  It is still possible to select MAIN and then de-select other stations, however they will still be active as this is also enforced in connect_to_lobby_button_pressed().  So the UI discrepancy is reduced but not totally eliminated.

415703b6d4e4d6be6535e0d62b08d6184702c6b2
